README.md
SAF
===
SAF(Simple Android Framework)是一个简单的android框架，它为开发Android app提供了基础性组件。  
SAF已经在多个项目中使用，包括今夜酒店特价app、锦江之星app、京东内部的多个app等等。这个项目第一次提交到google code是2012年的3月26号，我已经断断续续做了4年多了。2014年9月开始[frankswu](https://github.com/frankswu)加入跟我一起开发SAF。  
目前google code上的工程暂停维护，迁移到github上。它到了1.1.19版本，下一个版本肯定会是1.2.0。
目前已经增加了Rxjava作为依赖库,做了RxEventBus和RxImageLoader这两个模块,不过还不太成熟,不推荐生产中使用,下一个版本会进行改进。
遇到任何问题欢迎跟我的qq联系，qq：63067756, 玩得开心:)

我也觉得SAF越来越大了,怎么办呢? 好在SAF通过gradle进行构建项目,在命令行中输入gradle,即可一键生成多种jar包.

除了SAF自带的demo外，[魔窗的android demo app](https://github.com/magicwindow/mw-androidsdk-example)也使用了SAF，而且也是开源的

![](logo.png)

主要功能
-----------------------------------
* [SAFApp](https://github.com/fengzhizi715/SAF#safapp)
* [New Annotation](https://github.com/fengzhizi715/SAF#new-annotation-without-reflection)
* [Event Bus](https://github.com/fengzhizi715/SAF#event-bus)
* [Rest Client](https://github.com/fengzhizi715/SAF#rest-client)
* [Image Cache](https://github.com/fengzhizi715/SAF#image-cache)
* [Dependency Injection](https://github.com/fengzhizi715/SAF#dependency-injection)
* [Sqlite ORM](https://github.com/fengzhizi715/SAF#sqlite-orm)
* [Router](https://github.com/fengzhizi715/SAF#router)
* [Cache](https://github.com/fengzhizi715/SAF#cache)
* [Utils](https://github.com/fengzhizi715/SAF#utils)

SAFApp
===
SAFApp其实不能算是一个完整的模块，SAFApp继承了Application。增加了一个可作为缓存存放app全局变量的session，一个ImageLoader，一个记录Activity的List。

New annotation without reflection
===
新的注解，基于aspectj的AOP，而无需再使用耗费性能的反射.
不过,需要在build.gradle中配置一下aspectj


| 注解名称        | 作用          | 备注          |
| ------------- |:-------------:| :-------------:|
| @Async        |借助rxjava,异步执行app中的方法|       |
| @Cacheable    |Spring Cache风格的Cache注解,将结果放于缓存中|只适用于android4.0以后|
| @LogMethod    |将方法的入参和出参都打印出来,可以用于调试|       |
| @Prefs        |将方法的结果放入AppPrefs中|只适用于android4.0以后|
| @Safe         |可以安全地执行方法,而无需考虑是否会抛出运行时异常|       |
| @Trace        |用于追踪某个方法花费的时间,可以用于性能调优的评判|       |

@Async的使用方法:
<pre><code>
	@Async
	private void useAsync() {
		Log.e(TAG, " thread=" + Thread.currentThread().getId());
		Log.e(TAG, "ui thread=" + Looper.getMainLooper().getThread().getId());
	}
</pre></code>


@Cacheable的使用方法:

	@Cacheable(key = "user")
	private User initData() {

		User user = new User();
		user.userName = "tony";
		user.password = "123456";
		return user;
	}

这里的@Cacheable,实际上用到[Cache](https://github.com/fengzhizi715/SAF#cache),要获取Cache也很简单.

@Trace的使用方法:
<pre><code>
	@Trace
	@Async
	private void loadUser() {
		Log.e(TAG, " thread=" + Thread.currentThread().getId());
		Log.e(TAG, "ui thread=" + Looper.getMainLooper().getThread().getId());
		Cache cache = Cache.get(this);
		User user = (User) cache.getObject("user");
		Toast.makeText(MainActivity.this, SAFUtils.printObject(user), Toast.LENGTH_SHORT).show();
	}
</pre></code>
将@Trace和@Async两个注解结合使用,可以看到loadUser()方法花费的时间.
<pre><code>
05-18 14:31:31.229 21190-21190/app.magicwindow.cn.testsaf I/MainActivity: MainActivity=loadUser() take [1ms]<br>
05-18 14:31:31.231 21190-22033/app.magicwindow.cn.testsaf E/com.test.saf.activity.MainActivity:  thread=14876<br>
05-18 14:31:31.231 21190-22033/app.magicwindow.cn.testsaf E/com.test.saf.activity.MainActivity: ui thread=1<br>
</pre></code>

Event Bus
===
事件总线框架，类似于google guava、square otto的event bus。它是一种消息发布-订阅模式,它的工作机制类似于观察者模式，通过通知者去注册观察者，最后由通知者向观察者发布消息。

Event Bus解耦了asyncTask、handler、thread、broadcast等组件。使用Event bus可以轻松地跨多个Fragment进行通讯。

它用法很简单，在Activity或者Fragment中使用，其中event是一个简单的POJO<br />
<pre><code>
// 退出系统的事件
eventBus.post(new LogoutEvent());
</pre></code>

回调事件，同样在Activity或者Fragment中定义好。回调方法名可以随便定义，参数须要和event一一对应。并且在方法名前加上注解Subscribe

         /**
          * 退出整个app
          * @param event
          */
          @Subscribe
          public void onLogoutEvent(LogoutEvent event) {
          }
          
          
@Subscribe可以使用枚举<br />

         /**
          * 使用ThreadMode.BackgroundThread枚举，表示在后台线程运行，不在主线程中运行。
          * @param event
          */
          @Subscribe(ThreadMode.BackgroundThread)
          public void onBackendFresh(BackendFreshEvent event) {
          
          }
使用枚举BackgroundThread时，如果在回调方法中需要更新ui，则必须要配合handler使用。 在不使用枚举的情况下，@Subscribe会默认使用PostThread，表示回调方法会在主线程中运行。 如果在一个Activity中存在多个Fragment，并且在Activity或者在Fragment中存在订阅同一event的回调方法。如果发出event的请求时，这些回调方法都会起作用。


Rest Client
===
Rest Client模块提供了http的get、post、put、delete方法。这个模块还不是很完善，只是适应自身项目需要，未来会不断增加新的功能。 这个模块没有基于apache httpclient，完全基于jdk中的HttpURLConnection。

同步调用get方法：
<pre><code>
          RestClient client = RestClient.get(url);<p>
          String body = client.body();
</pre></code>

异步调用get方法：

          RestClient.get(url,new HttpResponseHandler(){
              
              public void onSuccess(String content) {
                // content为http请求成功后返回的response
              }
              
             @Override
			 public void onFail(RestException exception){
						
			  }
          });


同步调用post方法：post body内容为json

          RestClient client = RestClient.post(url);
          client.acceptJson().contentType("application/json", null);
          client.send(jsonString); // jsonString是已经由json对象转换成string类型
          String body = client.body();


异步调用post方法：post body内容为json

          RestClient.post(url,json,new HttpResponseHandler(){ // json对应的是fastjson的JSONObject对象
        
             public void onSuccess(String content) {
             }
                
             @Override
			 public void onFail(RestException exception){
						
			 }
        
           });


异步调用post方法：以form形式传递数据

          RestClient.post(urlString, map, new HttpResponseHandler(){

              @Override
              public void onSuccess(String content) {

              }
                
              @Override
			   public void onFail(RestException exception){
			    }
                                        
          });


Image Cache
===
图片缓存模块包括2级缓存，内存中的cache和sd卡上存放在文件中的cache。

图片缓存模块通过ImageLoader进行图片加载。 如果app中使用了SAFApp，则无须创建新的ImageLoader就可以使用。     
<pre><code>
          // 第一个参数是图片的url，第二个参数是ImageView对象，第三个参数是默认图片
          imageLoader.displayImage(url, imageView ,R.drawable.defalut_icon);
</pre></code>


Dependency Injection
===
Dependency Injection是依赖注入的意思，简称DI。

SAF中的DI包括以下几个方面：
* Inject View ：简化组件的查找注册，目前支持约定大于配置，如果代码中的组件名称跟layout中要注入的组件id相同，则无需写(id=R.id.xxxx)
* Inject Views：支持多个相同类型组件的注入
* Inject Service ：简化系统服务的注册，目前只支持android的系统服务
* Inject Extra ：简化2个Activity之间Extra传递
* InflateLayout ：简化布局填充时，组件的查找注册
* OnClick：简化各种组件的Click事件写法，其属性after可以指定一个方法在OnClick之后执行(一般用于埋点或者其共性操作)。
* OnItemClick：简化ListView的ItemView事件写法

Inject View
---
Inject View可以简化组件的查找注册，包括android自带的组件和自定义组件。在使用Inject View之前，我们会这样写代码

          public class MainActivity extends Activity {
                
                private ImageView imageView;
                
                @Override
                protected void onCreate(Bundle savedInstanceState) {
                  super.onCreate(savedInstanceState);
                  
                  setContentView(R.layout.activity_main);
                  imageView = (ImageView) findViewById(R.id.imageview);
                }
           }


在使用Inject View之后，会这样写代码

          public class MainActivity extends Activity {
                    
                @InjectView(id= R.id.imageview)
                private ImageView imageView;
                    
                @Override
                protected void onCreate(Bundle savedInstanceState) {
                   super.onCreate(savedInstanceState);
                      
                   setContentView(R.layout.activity_main);
                   Injector.injectInto(this);
                }
          }

约定大于配置的写法，无需写(id= R.id.imageview)

          public class MainActivity extends Activity {
                    
                @InjectView
                private ImageView imageview;
                    
                @Override
                protected void onCreate(Bundle savedInstanceState) {
                   super.onCreate(savedInstanceState);
                      
                   setContentView(R.layout.activity_main);
                   Injector.injectInto(this);
                }
          }

目前，@InjectView可用于Activity、Dialog、Fragment,Adapter(ViewHolder)中。在Activity和Dialog用法相似，在Fragment和Adapter中用法有一点区别。

          public class DemoFragment extends Fragment {

                   @InjectView(id=R.id.title)
                   private TextView titleView;

                   @InjectView(id=R.id.imageview)
                   private ImageView imageView;

                   @Override
                   public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
                          View v = inflater.inflate(R.layout.fragment_demo, container, false);

                          Injector.injectInto(this,v); // 和Activity使用的区别之处在这里
          
                          initViews();
                          initData();
          
                          return v;
                   }
          
                  ......
           }

Inject Views
---
          public class MainActivity extends Activity {
                    
                @InjectViews(ids={R.id.imageView1,R.id.imageView2})
                private List<ImageView> imageviews;
                    
                @Override
                protected void onCreate(Bundle savedInstanceState) {
                   super.onCreate(savedInstanceState);
                      
                   setContentView(R.layout.activity_main);
                   Injector.injectInto(this);
                }
          }


Inject Extra
---

         /**
          * MainActivity传递数据给SecondActivity
          * Intent i = new Intent(MainActivity.this,SecondActivity.class);                                               
          * i.putExtra("test", "saf");
          * i.putExtra("test_object", hello);
          * startActivity(i);
          * 在SecondActivity可以使用@InjectExtra注解
          *
          * @author Tony Shen
          *
          */
         public class SecondActivity extends Activity{

               @InjectExtra(key="test")
               private String testStr;
        
               @InjectExtra(key="test_object")
               private Hello hello;
        
               protected void onCreate(Bundle savedInstanceState) {
                   super.onCreate(savedInstanceState);
                
                   Injector.injectInto(this);
                   Log.i("++++++++++++","testStr="+testStr);
                   Log.i("++++++++++++","hello="+SAFUtil.printObject(hello)); // 该方法用于打印对象
              }
          }

InflateLayout
---

        /**
         * @author Tony Shen
         *
         */
         @InflateLayout(id=R.layout.my_view)
         public class MyView extends LinearLayout {

              @InjectView(id = R.id.textview1)
	          public TextView view1;
    
              @InjectView(id = R.id.textview2)
	          public TextView view2;
	
	         public MyView(Context context) {
		         super(context);
	         }
        }


在Activity、Fragment中的写法:
<pre><code> 	
         MyView myView = Injector.build(mContext, MyView.class);
</pre></code>


OnClick
---
@OnClick 可以在Activity、Fragment、Dialog、View中使用，支持多个组件绑定同一个方法。

     public class AddCommentFragment extends BaseFragment {
    
         @Override
         public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

             View v = inflater.inflate(R.layout.fragment_add_comment, container, false);

             Injector.injectInto(this, v);

             initView();

             return v;
        }
    
	    @OnClick(id={R.id.left_menu,R.id.btn_comment_cancel}, after = "pointClickView")
	    void clickLeftMenu() {
		    popBackStack();
	    }

      public void pointClickView(Method method, View view) {
        L.d("pointClickView");
        ....
      }
	
	    @OnClick(id=R.id.btn_comment_send)
	    void clickCommentSend() {
            if (StringHelper.isBlank(commentEdit.getText().toString())) {
               ToastUtil.showShort(mContext, R.string.the_comment_need_more_character);
            } else {
               AsyncTaskExecutor.executeAsyncTask(new AddCommentTask(showDialog(mContext)));
            }
	    }
	    
	    ....
    }

OnItemClick
---
<pre><code> 
	@OnItemClick(id=R.id.listview)
	void itemClickListView(AdapterView<?> parent, View view,
			int position, long id) {
		IncomeVO vo= (IncomeVO) listview.getItemAtPosition(position);				
		Intent i = new Intent(mContext,MyIncomeDetailActivity.class);
		Bundle bundle = new Bundle();
		bundle.putString("month", vo.month);
		bundle.putString("allIncome", Double.toString(vo.allIncome));
		i.putExtras(bundle);
		startActivity(i);
	}
</pre></code>

Sqlite ORM
===
顾名思义就是sqlite的orm框架，采用oop的方式简化对sqlite的操作。 </br>
<strong>注意：在android studio2.0以后如果使用InstantRun功能，并且在android 5.0以上手机上调试会报错。强烈建议在调试时关闭InstantRun功能，release版本的app不会受到任何影响。</strong>（参考：http://stackoverflow.com/questions/36572515/dexfile-in-2-0-versions-of-android-studio-and-gradle）</br>
首先需要在AndroidManifest.xml中配上一些参数

        <!-- 表示在com.example.testsaf.db这个package下的类都是db的domain，一个类对应db里的一张表-->
        <meta-data
            android:name="DOMAIN_PACKAGE"
            android:value="com.example.testsaf.db" />
        
        <!-- 表示db的名称-->
        <meta-data
            android:name="DB_NAME"
            android:value="testsaf.db" />
 
         <!-- 表示db的版本号-->
         <meta-data
            android:name="DB_VERSION"
            android:value="1" />


使用orm框架需要初始化DBManager，需要在Applicaion中完成。SAF中的SAFApp，没有初始化DBManager，如果需要使用SAFApp可以重写一个Application继承SAFApp，并初始化DBManager。

          /**
           * @author Tony Shen
           *
           */
           public class TestApp extends Application{

                @Override
                public void onCreate() {
                   super.onCreate();
                   DBManager.initialize(this);
                }
  
           }

db的domain使用是也是基于注解

          /**
           * 
           * 表示sqlite中autocomplete表的属性
           * @author Tony Shen
           * 
           */
          @Table(name="autocomplete")
          public class Autocomplete extends DBDomain{

              @Column(name="key_words",length=20,notNull=true)
              public String KEY_WORDS;
        
              @Column(name="key_type",length=20,notNull=true)
              public String KEY_TYPE;
        
              @Column(name="key_reference",length=80)
              public String KEY_REFERENCE;
          }


db的操作很简单

          Autocomplete auto = new Autocomplete();
          auto.KEY_TYPE = "1";
          auto.KEY_WORDS = "testtest";
          auto.save(); // 插入第一条记录

          Autocomplete auto2 = new Autocomplete();
          auto2.KEY_TYPE = "0";
          auto2.KEY_WORDS = "haha";
          auto2.save(); // 插入第二条记录

          Autocomplete auto3 = new Autocomplete().get(1); // 获取Autocomplete的第一条记录
          if (auto3!=null) {
               Log.i("+++++++++++++++","auto3.KEY_WORDS="+auto3.KEY_WORDS);
          } else {
               Log.i("+++++++++++++++","auto3 is null!");
          }


查询结果集
<pre><code>
List list = new Autocomplete().executeQuery("select * from autocomplete where KEY_WORDS = 'testtest'");
Log.i("+++++++++++++++","list.size()="+list.size());  // 根据sql条件查询
                
List list2 = new Autocomplete().executeQuery("select * from autocomplete where KEY_WORDS = ? and Id = ?","testtest","1");
Log.i("+++++++++++++++","list2.size()="+list2.size()); // 表示查询select * from autocomplete where KEY_WORDS = 'testtest' and Id = '1'
</pre></code> 


Router
===
类似于rails的router功能，可以实现app的应用内跳转,包括Activity之间、Fragment之间可以轻易实现相互跳转，并传递参数。 
使用Activity跳转必须在Application中做好router的映射。 我们会做这样的映射，表示从某个Activity跳转到另一个Activity需要传递user、password2个参数

          Router.getInstance().setContext(getApplicationContext()); // 这一步是必须的，用于初始化Router
          Router.getInstance().map("user/:user/password/:password", SecondActivity.class);


有时候，activity跳转还会有动画效果，那么我们可以这么做

          RouterOptions options = new RouterOptions();
          options.enterAnim = R.anim.slide_right_in;
          options.exitAnim = R.anim.slide_left_out;
          Router.getInstance().map("user/:user/password/:password", SecondActivity.class, options);


在Application中定义好映射，activity之间跳转只需在activity中写下如下的代码，即可跳转到相应的Activity，并传递参数
<pre><code>
           Router.getInstance().open("user/fengzhizi715/password/715");
</pre></code>

如果在跳转前需要先做判断，看看是否满足跳转的条件,doCheck()返回false表示不跳转，true表示进行跳转到下一个activity

          Router.getInstance().open("user/fengzhizi715/password/715",new RouterChecker(){

                 public boolean doCheck() {
                     return true;
                 }
          });


单独跳转到某个网页，调用系统电话，调用手机上的地图app打开地图等无须在Application中定义跳转映射。

          Router.getInstance().openURI("http://www.g.cn");

          Router.getInstance().openURI("tel://18662430000");

          Router.getInstance().openURI("geo:0,0?q=31,121");


Fragment之间的跳转也无须在Application中定义跳转映射。直接在某个Fragment写下如下的代码
<pre><code>
         Router.getInstance().openFragment(new FragmentOptions(getFragmentManager(),new Fragment2()), R.id.content_frame);
</pre></code>

当然在Fragment之间跳转可以传递参数
<pre><code>
Router.getInstance().openFragment("user/fengzhizi715/password/715",new FragmentOptions(getFragmentManager(),new Fragment2()), R.id.content_frame);
</pre></code>


Cache
===
这是一个通用的Cache,可以保存字符串、对象、JSON等等,操作起来十分简单,还可以设置缓存的过期时间.

保持缓存数据：
<pre><code>
      Cache cache = Cache.get(this);
      cache.put("key1", "test value");
      cache.put("key2", "test value", 10);//保存10秒钟，10秒后会过期
</pre></code>

获取缓存数据：
<pre><code>
      Cache cache = Cache.get(this);
      String value = cache.getString("key1");
</pre></code>

Utils
===
包含了很多常用的工具类，比如日期操作、字符串操作、SAFUtil里包含各种乱七八糟的常用类等等。




CHANGELOG.md
SAF
===
1.1.4之前的开发都是在google code上托管的，从1.1.4开始记录版本日志。

Version 1.1.19
---
2016-05-17
 *  针对android6.0的权限需求,imageloader做了修改
 *  新增NoEmptyHashMap
 *  优化RxImageLoader
 *  新增aspectj的支持, 增加了@LogMethod @Trace @Cacheable @Safe @Async @Prefs 等注解,这些注解无需使用反射
 *  新增RxAsyncTask

Version 1.1.18
---
2016-02-15
 *  修复底层的bug 优化代码
 *  项目结构由原先的eclipse转变成android studio
 *  增加工具类Preconditions
 *  增加RxEventBus RxImageloader,这2个模块还处于试验阶段,不推荐使用
 *  增加android5.0以后判断app是否处于前台的方法
 *  增加cache类
 *  增加SAFRecycleAdapter类
 *  去掉原先插件化的类,打算重新设计这个模块

Version 1.1.17
---
2015-08-28
 *  增加HttpRetryStrategy接口
 *  imageloader的线程数,由当前手机的cpu数量决定
 *  增加新型的反射工具类

Version 1.1.16
---
2015-08-04
 *  删除RestClient的BinaryResponseHandler
 *  RestClient增加response的header支持
 *  修复android L以后版本无法使用imageloader的bug

Version 1.1.15
---
2015-07-21
 *  增加下载模块
 *  修复删除图片缓存时,报的空指针bug

Version 1.1.14
---
2015-06-18
 *  增加AbstractParcelable
 *  精简SAFActivity、SAFFragmentActivity
 *  删除cn.salesuite.saf.net包,减少saf包的大小

Version 1.1.13
---
2015-05-06
 *  RestClient增加https的支持
 *  修复L的bug
 *  InjectExtra增加默认值
 *  AsyncTaskExecutor增加android.os.AsyncTask的支持

Version 1.1.12
---
2015-03-11
 *  增加OnClick的hocker接口，before和after方法，用于埋点
 *  增加OnItemClick的hocker接口，before和after方法，用于埋点
 *  增加工具方法
 *  增加RestClient的BinaryResponseHandler（曹亚民）
 *  增加全新的AsyncTask（曹亚民）

Version 1.1.11
---
2015-02-24
 *  增加工具类
 *  优化Router，Router增加便于埋点的接口
 *  修复ToastUtils的bug

Version 1.1.10
---
2014-12-24
 *  eventbus增加线程模式ScheduleBackgroundThread
 *  增加List的帮助类Lists
 *  BasePrefs中各个put方法，都包含save()方法，避免BasePrefs再调用save（）
 *  修复bug

Version 1.1.9
---
2014-11-12
 *  完善工具类命名规范

Version 1.1.8
---
2014-11-10
 *  完善SAFAdapter
 *  Injector增加Adapter的支持
 *  完善多个工具类

Version 1.1.7
---
2014-10-10
 *  优化ImageLoader
 *  增加@OnTouch
 *  增加SAFAdapter
 *  完善L、ToastUtil

Version 1.1.6
---
2014-09-17
 *  优化RestClient，增加重试机制
 *  增加@OnLongClick

Version 1.1.5
---
2014-09-11
 *  优化ImageLoader，disklrucache升级到2.0.3

Version 1.1.4
---
2014-09-06
 *  增加@InjectViews、@OnClick、@OnItemClick
 *  增加日志框架L
 *  优化@InjectView


src/cn/salesuite/saf/route/Router.java
/**
 * 
 */
package cn.salesuite.saf.route;

import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.util.LruCache;
import cn.salesuite.saf.route.RouterParameter.RouterOptions;

/**
 * 要使用Router功能时，必须在application中做好router的映射<br>
 * 从某个Activity跳转到SecondActivity需要传递user、password2个参数，需要做这样的映射
 * <pre>
 * <code>
 * Router.getInstance().map("user/:user/password/:password", SecondActivity.class);
 * </code>
 * </pre>
 * 跳转时可以增加动画效果
 * <pre>
 * <code>
 * RouterOptions options = new RouterOptions();
 * options.enterAnim = R.anim.slide_left_in;
 * options.exitAnim = R.anim.slide_right_out;
 * Router.getInstance().map("user/:user/password/:password", SecondActivity.class, options);
 * </code>
 * </pre>
 * Intent Router可以完成各个Intent之间的跳转，类似rails的router功能
 * @author Tony Shen
 *
 */
public class Router {
	
	public static final int DEFAULT_CACHE_SIZE = 1024;
	
	private Context context;
	private LruCache<String, RouterParameter> cachedRoutes = new LruCache<String, RouterParameter>(DEFAULT_CACHE_SIZE); // 缓存跳转的参数
	private final Map<String, RouterOptions> routes = new HashMap<String, RouterOptions>();                             // 存放Intent之间跳转的route
	
	private static final Router router = new Router();
	
	private Router() {
	}
	
	public static Router getInstance() {
		return router;
	}
	
	public void setContext(Context context) {
		this.context = context;
	}

	public Context getContext() {
		return context;
	}
	
	/**
	 * @param format 形如"user/:user/password/:password"这样的格式，其中user、password为参数名
	 * @param clazz
	 */
	public void map(String format, Class<? extends Activity> clazz) {
		this.map(format, clazz, null);
	}

	public void map(String format, Class<? extends Activity> clazz,RouterOptions options) {
		if (options == null) {
			options = new RouterOptions();
		}
		options.clazz = clazz;
		this.routes.put(format, options);
	}

	/**
	 * 跳转到网页，如下：
	 * <pre>
	 * <code>
	 * Router.getInstance().openURI("http://www.g.cn");
	 * </code>
	 * </pre>
	 * 
	 * 调用系统电话，如下：
	 * <pre>
	 * <code>
	 * Router.getInstance().openURI("tel://18662430000");
	 * </code>
	 * </pre>
	 * 
	 * 调用手机上的地图app，打开地图，如下：
	 * <pre>
	 * <code>
	 * Router.getInstance().openURI("geo:0,0?q=31,121");
	 * </code>
	 * </pre>
	 * @param url
	 */
	public void openURI(String url) {
		this.openURI(url,this.context);
	}
	
	public void openURI(String url,Context context) {
		this.openURI(url, context, null);
	}
	
	public void openURI(String url,Context context,Bundle extras) {
		openURI(url,context,extras,Intent.FLAG_ACTIVITY_NEW_TASK);
	}
	
	public void openURI(String url,Context context,Bundle extras, int flags) {
		openURI(url,context,extras,null,flags);
	}

	public void openURI(String url,Context context,Bundle extras, RouterPoint point, int flags) {
		if (context == null) {
			throw new RouterException("You need to supply a context for Router " + this.toString());
		}
		
		Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
		this.addFlagsToIntent(intent, context, flags);
		if (extras != null) {
			intent.putExtras(extras);
		}
		
		// frankswu : 在start之前增加相应的接口可以做类似埋点的工作
		if (point != null) {
			point.beforeRouter(url, extras);
		}
		
		context.startActivity(intent);
		
		if (point != null) {
			point.afterRouter(url, extras);
		}
	}
	
	/**
	 * 跳转到某个activity并传值
	 * <pre>
	 * <code>
	 * Router.getInstance().open("user/fengzhizi715/password/715");
	 * </code>
	 * </pre>
	 * @param url
	 */
	public void open(String url) {
		this.open(url,null);
	}
	
	/**
	 * 跳转到某个activity并传值,router跳转前的先判断是否满足跳转的条件,doCheck()返回false表示不跳转，true表示进行跳转到下一个activity
	 * <pre>
	 * <code>
	 * Router.getInstance().open("user/fengzhizi715/password/715",new RouterChecker(){
     *
	 *	 public boolean doCheck() {
	 *		 return false;
	 *	 }
	 * });
	 * </code>
	 * </pre>
	 * @param url
	 * @param checker
	 */
	public void open(String url,RouterChecker checker) {
		this.open(url,this.context,checker);
	}
	
	public void open(String url,Context context,RouterChecker checker) {
		this.open(url, context, null,checker);
	}
	
	public void open(String url,Context context,Bundle extras) {
		open(url,context,extras,Intent.FLAG_ACTIVITY_NEW_TASK,null,null); // 默认的跳转类型,将Activity放到一个新的Task中
	}
	
	public void open(String url,Context context,Bundle extras,RouterChecker checker) {
		open(url,context,extras,Intent.FLAG_ACTIVITY_NEW_TASK,checker,null); // 默认的跳转类型,将Activity放到一个新的Task中
	}
	
	public void open(String url,Context context,Bundle extras,int flags,RouterChecker checker, RouterPoint point) {
		if (context == null) {
			throw new RouterException("You need to supply a context for Router "+ this.toString());
		}
		
		if (checker!=null && !checker.doCheck()) {
			return;
		}
		
		RouterParameter param = parseUrl(url);
		RouterOptions options = param.routerOptions;
		
		Intent intent = this.parseRouterParameter(param);
		if (intent == null) {
			return;
		}
		if (extras != null) {
			intent.putExtras(extras);
		}
		this.addFlagsToIntent(intent, context, flags);
		
		// frankswu : 在start之前增加相应的接口可以做类似埋点的工作
		if (point != null) {
			point.beforeRouter(url, extras);
		}
		
		context.startActivity(intent);
		
		if (point != null) {
			point.afterRouter(url, extras);
		}
		
		if (options.enterAnim>0 && options.exitAnim>0) {
			((Activity)context).overridePendingTransition(options.enterAnim, options.exitAnim);
		}
	}
	
	public void openFragment(FragmentOptions fragmentOptions,int containerViewId) {
		if (!(fragmentOptions != null
				&& fragmentOptions.mFragmentInstnace != null
				&& fragmentOptions.fragmentManager != null))
			return;
		
		if (fragmentOptions.mArg!=null) {
			fragmentOptions.mFragmentInstnace.setArguments(fragmentOptions.mArg);
		}
		
		fragmentOptions.fragmentManager.beginTransaction().replace(containerViewId , fragmentOptions.mFragmentInstnace).addToBackStack(null).commit();
	}
	
	public void openFragment(String url,FragmentOptions fragmentOptions,int containerViewId,RouterPoint point) {
		if (!(fragmentOptions != null
				&& fragmentOptions.mFragmentInstnace != null
				&& fragmentOptions.fragmentManager != null))
			return;
		
		Fragment fragment = fragmentOptions.mFragmentInstnace;
		fragment = parseFragmentUrl(url,fragmentOptions);
		
		// TODO frankswu : 在fragment跳转之前增加相应的接口可以做类似埋点的工作
		if (point != null) {
			point.beforeRouter(url, fragmentOptions.mArg);
		}
		
		fragmentOptions.fragmentManager.beginTransaction().replace(containerViewId , fragment).addToBackStack(null).commit();
		
		if (point != null) {
			point.afterRouter(url, fragmentOptions.mArg);
		}
	}
	
	private Fragment parseFragmentUrl(String url, FragmentOptions fragmentOptions) {
		String[] givenParts = url.split("/");
		int length = givenParts.length;
		
		if (length > 0) {
			if (fragmentOptions.mArg==null) {
				fragmentOptions.mArg = new Bundle();
			}
			
			for (int i=0;i<length;i=i+2){
				fragmentOptions.mArg.putString(givenParts[i], givenParts[i+1]);
			}
			
			fragmentOptions.mFragmentInstnace.setArguments(fragmentOptions.mArg);
		}

		return fragmentOptions.mFragmentInstnace;
	}

	/**
	 * intent增加额外的flag
	 * @param intent
	 * @param context
	 * @param flags
	 */
	private void addFlagsToIntent(Intent intent, Context context,int flags) {
		intent.addFlags(flags);
	}

	private Intent parseRouterParameter(RouterParameter param) {
		RouterOptions options = param.routerOptions;
		Intent intent = new Intent();

		for (Entry<String, String> entry : param.openParams.entrySet()) {
			intent.putExtra(entry.getKey(), entry.getValue());
		}
		
		intent.setClass(context, options.clazz);
		
		return intent;
	}

	private RouterParameter parseUrl(String url) {
		if (this.cachedRoutes.get(url) != null) {
			return this.cachedRoutes.get(url);
		}

		String[] givenParts = url.split("/");

		RouterOptions openOptions = null;
		RouterParameter openParams = null;
		for (Entry<String, RouterOptions> entry : this.routes.entrySet()) {
			String routerUrl = entry.getKey();
			RouterOptions routerOptions = entry.getValue();
			String[] routerParts = routerUrl.split("/");

			if (routerParts.length != givenParts.length) {
				continue;
			}

			Map<String, String> givenParams = urlToParamsMap(givenParts, routerParts);
			if (givenParams == null) {
				continue;
			}

			openOptions = routerOptions;
			openParams = new RouterParameter();
			openParams.openParams = givenParams;
			openParams.routerOptions = routerOptions;
			break;
		}

		if (openOptions == null || openParams == null) {
			throw new RouterException("No route found for url " + url);
		}

		this.cachedRoutes.put(url, openParams);
		return openParams;
	}

	private Map<String, String> urlToParamsMap(String[] givenUrlSegments, String[] routerUrlSegments) {
		Map<String, String> formatParams = new HashMap<String, String>();
		int length = routerUrlSegments.length;
		for (int index = 0; index < length; index++) {
			String routerPart = routerUrlSegments[index];
			String givenPart = givenUrlSegments[index];

			if (routerPart.charAt(0) == ':') {
				String key = routerPart.substring(1, routerPart.length());
				formatParams.put(key, givenPart);
				continue;
			}

			if (!routerPart.equals(givenPart)) { // 奇数个参数进行参数名的匹配，如果不相等则退出
				return null;
			}
		}

		return formatParams;
	}
	
	/**
	 * 退出系统时，清空缓存数据
	 */
	public void clear() {
		cachedRoutes.evictAll();
	}
	
	/**
	 * 
	 * @author Tony Shen
	 *
	 */
	public interface RouterChecker {
		boolean doCheck(); //router跳转前的先判断是否满足跳转的条件,false表示不跳转，true表示进行跳转到下一个activity
	}

	/**
	 * 提供埋点的便利接口，支持openURI和openFragment方法
	 * @author frankswu
	 *
	 */
	public interface RouterPoint {

		void beforeRouter(String url,Bundle extras);

		void afterRouter(String url,Bundle extras);

	}
}
